[클린코드 리액트 2기 신형기] 장바구니 미션 Step 2

hyungkishin opened the pull request:
# Pull Request Template

## 1. 가이드

### 1-1. 배포


- [LMS](https://clean-shop-cart.netlify.app/)

```bash
# install
yarn

# run
yarn dev

# test
yarn test
```

## 2. 리뷰 요청 ✍️

### 2-1. 어렵거나 아쉬웠던 점
- bottom up 은 역시 초반에, 시작할것 

### 2-2. 나누고 싶은 고민

### 2-3. 중점적으로 리뷰받고 싶은 부분
- react-query 만으로 충분함을 알고있지만, 카드 주문 count 상태를 변경 하기위해서는 server 에서 가져온 데이터말고 
더미 상태가 필요하다. 전역이 아닌 부모가 가지고 있어도 될것 같으나, 현재 구조가 살짝 아쉽다. 
```
cart
 ㄴ leftSection
     ㄴ index
     ㄴ item 
 ㄴ rightSection
     ㄴ index 
```

### 2-4. 새로운 시도 혹은 미션을 통해 도전한 점
- 상태관리 라이브러리 를 도입한것은 개인적으론 좋았다. (뚜드러 맞기전까진)
- 숙련도가 필요하다

### 2-5. 도식화된 자료 혹은 작성된 이미지 형태의 설계 구조

## 3. 질문있어요 🙋

## 4. 요구 사항 ✅

> 예시를 참고하여 체크 박스를 활용해주세요 (삭제)

### 4-1. 필수 요구사항
- [x] `MSW`를 활용한 `API mocking`
# 장바구니
- [x] 해당 상품의 수량을 변경할 수 있다.
  - [x] 상품의 수량은 항상 1이상, 20이하여야 한다
    - [x] 상품의 수량이 1이면 상품 수량 감소할 수 없다.
    - [x] 상품의 수량이 20이면 상품 수량 증가할 수 없다.
  - [x] 해당 상품의 총 금액이 변경된다.
  - [x] 해당 상품이 체크되어있으면, 결제예상금액도 변경된다.
  - [x] 모두선택 버튼이 체크되면, 상품들이 모두 선택된다.
    - [x] 모두선택 버튼이 체크가 풀리면, 상품들의 선택이 모두 해제된다.
  - [x] 상품 삭제 버튼을 누르면, confirm 메시지가 보여진다.
    - [x] 확인을 누르면, 선택된 상품이 모두 삭제된다.
    - [x] 결제예상금액이 0원이 된다.
  - [x] 🗑 버튼을 누르면 confirm 메시지가 보여진다.
    - [x] 확인을 누르면, 해당 상품이 삭제된다.
  - [x] 체크된 상품 개수에 따라 주문하기 버튼 내부에 수량이 변경된다.
  - [x] 주문하기 버튼을 누르면, confirm 메시지가 보여진다.
    - [x] 확인을 누르면, 주문/결제 페이지로 이동한다.
    - [x] 확인을 누르면, 장바구니에서 선택된 상품들이 삭제된다.
    - [x] 확인을 누르면, 체크된 상품들을 데이터베이스에서 제거한다.
  - [x] 주문할 상품이 0개이면 버튼이 비활성화된다.

### 4-2. 선택 요구사항 (심화)

- [ ] 반응형 레이아웃을 구현한다.

Commit:
feat: css 추가

Commit:
feat: cra 초기 setting

Commit:
feat: MSW lib 추가

Commit:
docs: 요구사항 체크

Commit:
feat: Header 컴포넌트 분리 및 with stitches/react

Commit:
feat: ItemList styling

Commit:
feat: nav styling

Commit:
feat: dbjson server 추가

Commit:
feat: axios 도입  productsPage Rendering

Commit:
feat: router 생성 ProductPage 연결

Commit:
feat: reset Css 추가

Commit:
feat: router 정리

Commit:
docs: 요구사항 체크

Commit:
feat: Nav Router 적용

Commit:
feat: icon 위치 변경, project default path 설정

Commit:
refactor: project import path 전체 수정

Commit:
refactor: 상품 목록 페이지 리펙토링 및 api 를 2번 call 하는 이유 확인후 store 생성준비

Commit:
refactor: 상품 목록 페이지 리펙토링

Commit:
feat: store > context, reducer 분리

Commit:
feat: server.js 를 바라보게 package.json 수정

Commit:
feat: css 민족 대이동 (도메인에 맞는 컴포넌트끼리)

Commit:
refactor: icon 경로 수정

Commit:
feat: carts 컴포넌트 추가 및 api 연동

Commit:
docs: 요구사항 체크

Commit:
feat: MSW 를 활용한 API mocking test

Commit:
docs: 요구사항 체크

Commit:
feat: route 정리

Commit:
docs: README update

Commit:
docs: 요구사항 체크

Commit:
refactor: mocks api > cart, product return data naming 변경

Commit:
refactor: 미사용 import 정리

Commit:
refactor: fileName 에 공백 제거

Commit:
refactor: actionType 네이밍 변경

Commit:
refactor: msw mocking, App.tsx 에 적용 및 코드리뷰 반영

Commit:
delete: 불필요 env 파일 제거

Commit:
refactor: global Header 수정

장바구니, 주문목록 선택시 페이지 랜딩

Commit:
feat: react-query, recoil 도입

사유: context를 현재 로그인한 유저, 테마, 선호하는 언어 등의 데이터 공유 시에 사용할 것을 추천
본 프로젝트에선 부적합하다고 판단.
물론 쓸라면 쓸순있겠지만 왜 사람들이 그렇게 recoil 을 찬양할까 비교글을 보아하니 단순하고 + 데이터를 atom 단위로 보는것과 여러가지 장점이 존재한다.
반면 현재 context-api 는 Provider와 컴포넌트 트리 노드 사이에 강한 커플링이 생겨버려서 쓸만큼 써봤고 지겨웠고. recoil 로 탈출하려고 한다.
react-query 는 서버 상태 가져오기, 캐싱, 동기화 및 업데이트를 쉽게 만들어주기 때문에 loading spinner 과 error 상태를 좀더 쉽게 뽑아먹을수 있다는 장점과
제일 중요한 캐싱 ... 동기화 이부분을 사용해보려고 한다.

Commit:
feat: react-query, recoil 적용

Commit:
chore: 파일 naming 변경 및, product data fetching querycache 1분 설정

Commit:
feat: react-query devtools 추가

Commit:
refactor: pageContent useEffect 내부 로직으로 이동

Commit:
refactor: json server start 구동 명령어로 merge

Commit:
feat: orderTest 및 hooks 작성

Commit:
feat: header -> global Header 이름 변경

Commit:
refactor: cart.ts 타이핑

Commit:
feat: orderList Page, DetailPage 작성

Commit:
refactor: common/ui Header, icon 공통화

Commit:
feat: modal component 추가 및 공통화

Commit:
refactor: common/ui, modal 분리

Commit:
feat: step-1 요구사항 체크 및, 상품상세 장바구니 선택시 이동

Commit:
refactor: 미사용 html 이동

Commit:
refactor: 코드리뷰 반영

Commit:
feat: 장바구니 추가 api, page 반영 및 mocks api 반영

Commit:
feat: 장바구니 기능 적용: [상품 갯수 증가, 감소],[모두선택]

Commit:
refactor: 모두 체크 로직 이동

Commit:
refactor: 장바구니 삭제 기능추가

Commit:
refactor: 장바구니 목록 전체 삭제 기능 구현 및 useCart 내부 장바구니 추가, 업데이트 hook namming 변경

Commit:
feat: 무한 스크롤 적용 -> product List

Commit:
feat: 주문 정보 Page 작성 및 store 연동

Commit:
refactor: tempState cart -> order hooks 이동 및 불필요 상태 제거

Commit:
fix: 장바구니 선택후 우측 상단의 주문목록 이동시 상태가 비워지지 않는 버그 수정

hyungkishin committed:
fix: 원격 branch conflict 해소

hyungkishin committed:
docs: 요구사항 체크

woobottle commented on src/page/cart/index.tsx:
```diff
@@ -19,15 +19,15 @@ const CartContent = () => {
   }
 
   return (
     <section className="cart-section">
```
cart List에서 받아온 데이터들을 활용하지 않는것 처럼 보이는데 의도된 것일까요??

hyungkishin replied:
대격변을 준비중입니다 ㅋㅋㅋㅋ

woobottle commented on src/page/product/list/item.tsx:
```diff
@@ -23,25 +26,32 @@ const Item = ({ item }: ItemProps) => {
   };
 
-  const handleAddCart = (e: React.MouseEvent<HTMLImageElement, MouseEvent>) => {
+  const HandleAddCart = (
```
함수명에 파스칼 케이스 쓰신 이유가 궁금해요!
위의 네이밍 규칙과는 다른것 같아서요

woobottle commented on src/router/index.tsx:
```diff
@@ -4,7 +4,7 @@ import Home from "page/home";
 import CartPage from "page/cart";
 import Detail from "page/product/detail";
-import OrdeList from "page/order/list";
+import OrderList from "page/order/already";
```
already로 네이밍 하신 이유가 궁금해요!!

hyungkishin replied:
저부분의 네이밍도 변경이 필요합니다 카트의 상황과 주문의 상황이 엄연히 다른데 잘못구성한듯하네요 :) 
이부분 리펙토링 하고 pr 올릴께요 :) !!

woobottle commented on src/page/product/list/index.tsx:
```diff
@@ -1,41 +1,52 @@
 
-  const { data, isLoading, isError } = useProductList();
-  const [products, setProducts] = useRecoilState(productListState);
+  const { isLoading: isScrollLoading, setIsLoading: setScrollLoading } =
```
1. 스크롤 시 데이터를 추가로 요청하는 부분은 보이지 않는것 같은데 추가로 구현 계획이실지 알고 싶어요!
2. reactQuery를 쓰시고 계시니까 useInfiniteScroll을 활용해보시면 어떠실까요?

woobottle commented on src/page/product/detail/index.tsx:
```diff
@@ -9,38 +9,75 @@ import { useSelectedProduct } from "hooks/product";
+    return <div>상품이 없습니다.</div>;
+  }
+
+  const HandleAddCart = (
```
이 친구도 파스칼 케이스가 의도된 것일지 궁금해요!

woobottle commented on src/page/product/detail/index.tsx:
```diff
@@ -9,38 +9,75 @@ import { useSelectedProduct } from "hooks/product";
-  
-  if (!selectedProduct) {
-    return <div>no Have Detail Item</div>;
+  const [product, setProduct] = useState<ProductItem>({} as ProductItem);
```
product상태를 따로 관리해야하는 상황이 궁금해요!

woobottle commented on src/page/cart/left-section/item.tsx:
```diff
@@ -11,25 +66,49 @@ const Item = ({ item }: { item: CartItem }) => {
-              <button className="number-input-button">▼</button>
+              <button
+                className="number-input-button"
+                onClick={() => setQuantity(quantity + 1)}
```
더 설명이 가능한 네이밍을 가진 함수를 위치시키면 어떻게 생각하실지 궁금해요!

addQuantity = () => {setQuantity(quantity + 1)}

onClick={addQuantity}

woobottle commented on src/page/cart/left-section/item.tsx:
```diff
@@ -11,25 +66,49 @@ const Item = ({ item }: { item: CartItem }) => {
+            className="cart-trash-svg"
+            src={trash}
+            alt="삭제"
+            onClick={() => handleDeleteCartItem()}
```
이러한 경우 handleDeleteCartItem만 반환해도 될 것 같은데 어떻게 생각하실지 궁금해요!!
onClick = {handleDeleteCartItem}

woobottle commented on src/page/cart/left-section/index.tsx:
```diff
@@ -1,10 +1,58 @@
+        "확인 버튼을 누르면 정말루 다 지워집니다. 그래도 삭제 가시겠습니까?",
+      onConfirm: () =>
+        // eslint-disable-next-line array-callback-return
+        carts.map((item: CartItem) => {
```
map은 변수를 반환하게 되는데 이 코드는 배열을 순회하기 위한 코드였을까요??

woobottle commented on src/page/cart/left-section/index.tsx:
```diff
@@ -1,10 +1,58 @@
+
+  const handleDeleteAll = () => {
+    handleModal({
+      title: "고갱님",
```
ㅋㅋㅋㅋㅋ

hyungkishin replied:
ㅋㅋㅋㅋㅋ

woobottle commented on src/page/cart/left-section/index.tsx:
```diff
@@ -1,10 +1,58 @@
+  useEffect(() => {
+    addTempAllCart(checked, carts.map((v: any) => v.product));
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
훅안에서 훅을 호출하는 것은 리액트에서 사용하는 훅의 규칙과 맞지 않는 것으로 알고 있는데요

어떤한 의도 였을지 궁금해요!!

woobottle commented on src/hooks/useScrollPagination.tsx:
```diff
@@ -0,0 +1,30 @@
+      if (scrollTop + clientHeight + threshold >= scrollHeight && !isLoading) {
+        setIsLoading(true);
+        onLoadMore();
+      }
```
혹시 https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API 를 활용하는 것은 어떻게 생각하실지 궁금해요!

woobottle commented on src/page/cart/left-section/item.tsx:
```diff
@@ -1,8 +1,63 @@
+  const [quantity, setQuantity] = useState(1);
+  const [singleChecked, setSingleChecked] = useState(false);
+
+  useEffect(() => {
```
이 부분은 페어프로그래밍때 같이 논의해보시죠~~

woobottle commented on src/page/cart/right-section/index.tsx:
```diff
@@ -1,4 +1,60 @@
+    // TODO : 주문하기 버튼 클릭시 장바구니에서 삭제 refactor
+    mutate(order);
+    // eslint-disable-next-line array-callback-return
+    return order.map((item) => {
```
이 함수에서 배열을 반환하고 있는 이유가 궁금해요!!

woobottle reviewed (changes requested):
프로젝트 데모 링크를 확인했을때 계속 데이터를 불러오다 실패해서 확인한번 부탁드릴게요!(아래 사진 첨부)
데모 링크 확인해주시고 우선 질문 드린 부분들 확인해주시면 그에 따라 추가적으로 질문 드릴 수 있도록 할게요~~

고생하셨습니다~~


<img width="851" alt="image" src="https://user-images.githubusercontent.com/50283326/228843470-72804a08-f9ea-45e2-ae0d-4e8dbf9e91c3.png">

hyungkishin committed:
ing: 중간 커밋 (cart) -> cart state 위치변경

hyungkishin committed:
refactor:
type -> global.d.ts 로 변경
hooks 리펙토링 -> carts, orders
checkBox , Divider 컴포넌트로 분리

hyungkishin committed:
refactor: orderListPage 리펙토링

hyungkishin committed:
refactor: mocks msw

hyungkishin committed:
fix: cart delete error fix

hyungkishin committed:
refactor: 페어코딩 적용

hyungkishin committed:
refactor: 코드리뷰 반영: recoil 제거 -> 커스텀 훅으로만 상태관리 방향

hyungkishin committed:
refactor: recoil 전체 삭제

hyungkishin committed:
refactor: item 제거후 unRerender fix
